core/sail/inferencer/src/main/java/org/eclipse/rdf4j/sail/inferencer/fc/FastRdfsForwardChainingSail.java

```diff
@@ -69,6 +73,76 @@ void clearInferenceTables() {
         calculatedDomain = new HashMap<>();
     }
 
+    void readLock(FastRdfsForwardChainingSailConnection connection) {
+
+        if (starveReads.get() > 0) {
+            System.err.println("starveReads");
+        }
+
+        while (starveReads.get() > 0) {
+            Thread.yield();
+        }
+
+        if (connection.lockStamp != 0) {
+            throw new IllegalStateException("Connection already have a lock!");
+        }
+        connection.lockStamp = readWriteLock.readLock();
+        System.err.println("readLock: " + connection.lockStamp);
+
+    }
+
+    void releaseLock(FastRdfsForwardChainingSailConnection connection) {
+        readWriteLock.unlock(connection.lockStamp);
+        System.err.println("Released lock: " + connection.lockStamp);
+
+        connection.lockStamp = 0;
+    }
+
+    AtomicInteger starveReads = new AtomicInteger(0);
+
+    void upgradeLock(FastRdfsForwardChainingSailConnection connection) {
+
+//        System.err.println("Attempt writelock: "+connection.lockStamp);
+
+        starveReads.incrementAndGet();
+
+        try {
+            for (int i = 0; i < random.nextInt(100) + 50; i++) {
```

Comment:
Why are you doing this rather complex looping (with a random value) at all? Why not simply issue a blocking call to obtain the write lock?